File: solutions/go/list-ops/1/list_ops.go
package listops

// IntList is an abstraction of a list of integers which we can define methods on
type IntList []int

func (s IntList) Foldl(fn func(int, int) int, initial int) int {
	r := initial
	for _, x := range s {
		r = fn(r, x)
	}
	return r
}

func (s IntList) Foldr(fn func(int, int) int, initial int) int {
	r := initial
	for i := len(s) - 1; i >= 0; i-- {
		r = fn(s[i], r)
	}
	return r
}

func (s IntList) Filter(fn func(int) bool) IntList {
	r := IntList{}
	for _, x := range s {
		if fn(x) {
			r = append(r, x)
		}
	}
	return r
}

func (s IntList) Length() int {
	return len(s)
}

func (s IntList) Map(fn func(int) int) IntList {
	r := IntList{}
	for _, x := range s {
		r = append(r, fn(x))
	}
	return r
}

func (s IntList) Reverse() IntList {
	r := IntList{}
	for i := len(s) - 1; i >= 0; i-- {
		r = append(r, s[i])
	}
	return r
}

func (s IntList) Append(lst IntList) IntList {
	return append(s, lst...)
}

func (s IntList) Concat(lists []IntList) IntList {
	r := s
	for _, list := range lists {
		r = append(r, list...)
	}
	return r
}

File: solutions/haskell/list-ops/1/src/ListOps.hs
{-# LANGUAGE BangPatterns #-}

module ListOps (
    length,
    reverse,
    map,
    filter,
    foldr,
    foldl',
    (++),
    concat,
) where

import Prelude hiding (concat, filter, foldr, length, map, reverse, (++))

foldl' :: (b -> a -> b) -> b -> [a] -> b
foldl' f z [] = z
foldl' f z (x : xs) = let !y = f z x in foldl' f y xs

foldr :: (a -> b -> b) -> b -> [a] -> b
foldr f z [] = z
foldr f z (x : xs) = f x (foldr f z xs)

length :: [a] -> Int
length = foldl' (\acc _ -> acc + 1) 0

reverse :: [a] -> [a]
reverse = foldl' (flip (:)) []

map :: (a -> b) -> [a] -> [b]
map f = foldr ((:) . f) []

filter :: (a -> Bool) -> [a] -> [a]
filter p = foldr (\x acc -> if p x then x : acc else acc) []

(++) :: [a] -> [a] -> [a]
xs ++ ys = foldr (:) ys xs

concat :: [[a]] -> [a]
concat = foldr (++) []
